README.md
# kontrol

[![Build Status](https://travis-ci.com/kalisio/kontrol.png?branch=master)](https://travis-ci.com/kalisio/kontrol)

**A controller that monitor and fix your Docker-based microservices infrastructure.**

kontrol is a very lightweight application you can run as a Docker container or a Docker Swarm service. It runs requests against a set of resources (containers or services) to determine whether those resources are operating normally (a.k.a. health check). Whenever a health check fails, it can launch commands against the Docker API in order to heal the infrastructure.

kontrol includes [Slack](https://slack.com/) notifications using [Incoming Webhook](https://api.slack.com/messaging/webhooks). Insert the webhook URL into your environment and enable the Slack integration. Anytime a healthcheck fails, the channel that you specified on Slack will receive a message.

## Why kontrol ?

Docker provides a built-in [healthcheck](https://docs.docker.com/engine/reference/builder/#healthcheck) for containers but it can be too much limited in some use cases. First, you will need to use a third-party tool anyway if you'd like to be notified on Slack about healthcheck failures. Second, the healing operation of Docker only consists in restarting the faulty container itself. In a complex microservices scenario this might not be sufficient as some containers depend on others to be reachable.

For instance, let imagine you have a service A using a service B, service A and B could be seen as healthy from the Docker point of view while the service B is not reachable from the service A for some reason (eg network failure, timeout, etc.). In that case it might be interesting to exhibit a healthcheck endpoint in service A providing the status of service B from the point of view of A. When that healthcheck fails in service A you should be notified and possibly the service B automatically restarted.

If you are looking for a solution to publish error logs to Slack you should better have a look to our [Slack adapter](https://github.com/kalisio/logspout-slack) for [logspout](https://github.com/gliderlabs/logspout).

## Configuring

Most of the configuration options come from [got](https://github.com/sindresorhus/got) and [dockerode](https://github.com/apocas/dockerode) powering kontrol. The exported object in the `config.js` should be structured like this:
* `docker`: the options used to initialize dockerode
* `jobs`: a map of healthcheck and notification/heal tasks to be performed, for each task identified by its key
  * `cron`: the [CRON pattern](https://github.com/kelektiv/node-cron) to schedule it
  * `delay`: the delay in seconds before scheduling the task
  * `notify`: optional function with the `error` object as input when the healthcheck has failed or none when it is back to healthy after a failure (returns the Slack message payload to be sent for notification)
  * `heal`: optional function with the `docker` object (i.e. dockerode instance) and `_` (lodash instance) as input and performing Docker commands to heal the infrastructure
  * all other options are sent to got instance for the healthcheck request

Here are the environment variables you can use to customize the behaviour:
| Variable  | Description | Defaults |
|-----------| ------------| ------------|
| `CONFIG_FILEPATH` | your configuration file path | `config.js` |
| `PORT` | the server port | `8080` |
| `SLACK_WEBHOOK_URL` | your Slack webhook URL |  |

## Example

The default [`config.js`](./config.js) is a great example to start from. It basically checks the kontrol container itself and restart it on failure, like would Docker do. When under test mode, the kontrol container will randomly fail with a `500` status code, so that the container will restart as long as you don't kill it manually. To build and launch the example execute the following commands:
```
SLACK_WEBHOOK_URL=https://hooks.slack.com/services/xxx
docker-compose build
docker-compose up
```

## Building

### Manual build 

You can build the image with the following command:

```bash
docker build -t <your-image-name> .
```

### Automatic build using Travis CI

This project is configured to use Travis to build and push the image on the Kalisio's Docker Hub.
The built image is tagged using the `version` property in the `package.json` file.

To enable Travis to do the job, you must define the following variable in the corresponding Travis project:

| Variable  | Description |
|-----------| ------------|
| `DOCKER_USER` | your username |
| `DOCKER_PASSWORD` | your password |

## Deploying

This image is designed to be deployed using the [Kargo](https://kalisio.github.io/kargo/) project.

Check out the [compose file](https://github.com/kalisio/kargo/blob/master/deploy/kontrol.yml) to have an overview on how the container is deployed.

## Contributing

Please read the [Contributing file](./.github/CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

## Versioning

We use [SemVer](https://semver.org/) for versioning. For the versions available, see the tags on this repository.

When releasing a patch, minor or major version, i.e. the following tasks have to be done:
- increase the package version number in the package.json file
- create a tag accordingly in the git repository and push it

The command `npm run release:<type>`, where  `<type>` is either `patch`, `minor` or `major`, will do the job for you ! 

## Authors

This project is sponsored by 

![Kalisio](https://s3.eu-central-1.amazonaws.com/kalisioscope/kalisio/kalisio-logo-black-256x84.png)

## License

This project is licensed under the MIT License - see the [license file](./LICENSE.md) for details


CHANGELOG.md
## 1.0.0 (2021-05-16)



config.js
module.exports = {
  docker: {},
  jobs: {
    kontrol: {
      cron: '*/30 * * * * *',
      url: 'http://localhost:8080/healthcheck',
      method: 'GET',
      headers: {},
      searchParams: {},
      retry: 1,
      timeout: 10000,
      notify: (error) => ({
         attachments: [{
          color: (error ? 'danger' : 'good'),
          mrkdwn_in: ['text'],
          text: (error ? `*kontrol healthcheck failed*\n${error}\nNow healing by restarting`
                       : `*kontrol healthcheck succeeded*`)
        }]
      }),
      heal: async (docker, _) => {
        const containers = await docker.listContainers()
        let container = _.find(containers, (container => container.Image === 'kalisio/kontrol'))
        container = await docker.getContainer(container.Id)
        await container.restart()
      }
    }
  }
}


index.js
#!/usr/bin/env node
const path = require('path')
const cors = require('cors')
const express = require('express')
const cron = require('cron')
const request = require('got')
const _ = require('lodash')
const Docker = require('dockerode')

const port = process.env.PORT || 8080
const config = require(process.env.CONFIG_FILEPATH || path.join(__dirname, 'config.js'))

let server
let jobs = {}
const docker = new Docker(config.docker)

async function publishToSlack (task, body) {
  if (!process.env.SLACK_WEBHOOK_URL) return
  try {
    await request({
      url: process.env.SLACK_WEBHOOK_URL,
      method: 'POST',
      body: JSON.stringify(body)
    })
  } catch (error) {
    // Allowed to fail to make healthcheck robust
    console.error(`Notification for task ${task} failed`, error)
  }
}

// Setup and launch server
function serve() {
  const app = express()
  app.use(cors()) // add CORS headers -- required

  // Healthcheck
  app.get('/healthcheck', (req, res) => {
    res.set('Content-Type', 'application/json')
    let response = {
      isRunning: true
    }
    // Expose all health states
    _.forOwn(jobs, (value, key) => {
      response[key] = value.health
    })

    // Report random errors when under test
    if (process.env.UNDER_TEST) return res.status(Math.random() > 0.5 ? 200 : 500).json(response)
    else return res.status(200).json(response)
  })
  // Start the server
  server = app.listen(port, () => {
    console.log('kontrol server listening at %d', port)
  })
}

// Plan all jobs
function plan() {
  for (const [key, jobOptions] of Object.entries(config.jobs)) {
    console.log(`Registering task ${key}`)
    const cronJob = new cron.CronJob(jobOptions.cron, async () => {
      let job = jobs[key]
      if (job.isRunning) {
        console.log(`Skipping scheduled task ${key} as previous one is not yet finished`)
        return
      }
      job.isRunning = true
      console.log(`Executing scheduled task ${key}`)
      job.previousHealth = job.health
      try {
        const response = await request(_.omit(jobOptions, ['cron', 'notify', 'heal', 'delay']))
        console.log(`Scheduled task ${key} response:`, response.statusCode, response.statusMessage)
        job.health = {
          statusCode: response.statusCode,
          statusMessage: response.statusMessage
        }
        // Notify when going back to normal
        if (jobOptions.notify && _.get(job, 'previousHealth.error')) {
          console.log(`Notifying healthy state for task ${key}`)
          try {
            const body = await jobOptions.notify()
            await publishToSlack(key, body)
          } catch (error) {
            console.error(`Notification for task ${key} failed`, error)
          }
        }
      } catch (error) {
        console.error(`Scheduled task ${key} error`, error)
        job.health = {
          error
        }
        // Notify when entering error state
        if (jobOptions.notify && !_.get(job, 'previousHealth.error')) {
          console.log(`Notifying healthcheck failure for task ${key}`)
          try {
            const body = await jobOptions.notify(error)
            await publishToSlack(key, body)
          } catch (error) {
            console.error(`Notification for task ${key} failed`, error)
          }
        }
        if (jobOptions.heal) {
          console.log(`Performing healing for task ${key}`)
          try {
            await jobOptions.heal(docker, _)
          } catch (error) {
            console.error(`Healing for task ${key} failed`, error)
            try {
              await publishToSlack(key, {
                attachments: [{
                  color: 'danger',
                  mrkdwn_in: ['text'],
                  text: `*Healing for task ${key} failed*\n${error}`
                }]
              })
            } catch (error) {}
          }
        }
      }
      job.isRunning = false
    })
    jobs[key] = { cronJob, health: {} }
    // Start immediately or after a given delay
    const delay = jobOptions.delay || 0
    setTimeout(() => cronJob.start(), delay)
  }
}

process.on('SIGTERM', () => {
  console.info('SIGTERM signal received, closing now.')
  _.forOwn(jobs, value => {
    value.cronJob.stop()
  })
  server.close(() => {})
})

async function run () {
  plan()
  serve()
}

run()
